File: 46. Permutations.cpp
class Solution {
public:
    // recursive: 7.25%, ok
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> ret;
        if(nums.empty()) return ret;
        if(nums.size() == 1){
            ret.push_back(vector<int>(1, nums[0]));
            return ret;
        }
        for(int i = 0; i < nums.size(); i++){
            vector<int> rest;
            for(int j = 0; j < nums.size(); j++)
                if(j != i) rest.push_back(nums[j]);
            vector<vector<int>> remain = permute(rest);
            for(int j = 0; j < remain.size(); j++){
                remain[j].insert(remain[j].begin(), nums[i]);
                ret.push_back(remain[j]);
            }
        }
        return ret;
    }
};

File: 46. Permutations.py
class Solution(object):
    
    # # back tracking, dfs: O(N!) time, O(1) space
    # def permute(self, nums):
    #     """
    #     :type nums: List[int]
    #     :rtype: List[List[int]]
    #     """
    #     def dfs(nums, curr, ret):
    #         if len(curr) == len(nums):
    #             ret.append(curr)
    #         else:
    #             for j in xrange(len(nums)):
    #                 if nums[j] not in curr:
    #                     dfs(nums, curr + [nums[j]], ret)
    #     ret = []
    #     dfs(nums, [], ret)
    #     return ret
    
    # back tracking, dfs: O(N!) time, O(1) space
    def permute(self, nums):
        
        def dfs(nums, curr, i, ret):
            if i == len(nums):
                ret.append(curr)
            else:
                for j in xrange(i, len(nums)):
                    nums[i], nums[j] = nums[j], nums[i]
                    dfs(nums, curr + [nums[i]], i + 1, ret)
                    nums[j], nums[i] = nums[i], nums[j]

        ret = []
        dfs(nums, [], 0, ret)
        return ret
        